Hi. Ok, maybe that is not helping code readability.
I tried casting to `int`, but it didn't fix the warning, see:
~~~
rodrigo@fedora base64]$ gcc main.c base64.c -o a -W
base64.c: In function ‘base64_decode’:
base64.c:134:36: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  134 |   if ((int)in[i] < 0 || (int)in[i] > 127) {
      |                                    ^
[rodrigo@fedora base64]$ gcc main.c base64.c -o a -W -funsigned-char
base64.c: In function ‘base64_decode’:
base64.c:134:18: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  134 |   if ((int)in[i] < 0 || (int)in[i] > 127) {
      |                  ^
~~~

I could copy the value to a local integer before comparing, but I don't know how that would be better than checking the MSB. What do you think about creating this other variable?

--- a/base64.c
+++ b/base64.c
@@ -3,6 +3,8 @@
 #include "base64.h"
 
 #define BASE64_PAD '='
+#define BASE64DE_FIRST '+'
+#define BASE64DE_LAST 'z'
 
 /* BASE 64 encode table */
 static const char base64en[] = {
@@ -131,7 +133,7 @@ base64_decode(const char *in, unsigned int inlen, unsigned char *out)
 		if (in[i] == BASE64_PAD) {
 			break;
 		}
-		if ((in[i] & 0x80) != 0) {
+		if (in[i] < BASE64DE_FIRST || in[i] > BASE64DE_LAST) {
 			return 0;
 		}
 

